redash/models.py

```diff
@@ -1034,7 +1012,134 @@ def record(cls, event):
+            'name': self.name,
+            'type': self.type,
+        }
+
+        if all:
+            d['options'] = self.configuration
+            d['groups'] = self.groups
+
+        if with_permissions:
+            d['view_only'] = self.notification_destination_groups.view_only
+
+        return d
+
+    def __unicode__(self):
+        return self.name
+
+    @classmethod
+    def create_with_group(cls, *args, **kwargs):
+        notification_destination = cls.create(*args, **kwargs)
+        NotificationDestinationGroup.create(
+            notification_destination=notification_destination, 
+            group=notification_destination.org.default_group
+        )
+        return notification_destination
+
+    @property
+    def configuration(self):
+        configuration = json.loads(self.options)
+        schema = self.destination.configuration_schema()
+        for prop in schema.get('secret', []):
+            if prop in configuration and configuration[prop]:
+                configuration[prop] = self.SECRET_PLACEHOLDER
+
+        return configuration
+
+    def replace_secret_placeholders(self, configuration):
+        current_configuration = json.loads(self.options)
+        schema = self.destination.configuration_schema()
+        for prop in schema.get('secret', []):
+            if prop in configuration and configuration[prop] == self.SECRET_PLACEHOLDER:
+                configuration[prop] = current_configuration[prop]
+
+    def add_group(self, group, view_only=False):
+        ndg = NotificationDestinationGroup.create(group=group, notification_destination=self, view_only=view_only)
+        setattr(self, 'notification_destination_groups', ndg)
+
+    def remove_group(self, group):
+        NotificationDestinationGroup.delete().where(NotificationDestinationGroup.group==group, NotificationDestinationGroup.notification_destination==self).execute()
+
+    def update_group_permission(self, group, view_only):
+        ndg = NotificationDestinationGroup.get(NotificationDestinationGroup.group==group, NotificationDestinationGroup.notification_destination==self)
+        ndg.view_only = view_only
+        ndg.save()
+        setattr(self, 'notification_destination_groups', ndg)
+
+    @property
+    def destination(self):
+        return get_destination(self.type, self.options)
+
+    @classmethod
+    def all(cls, org, groups=None):
+        notification_destinations = cls.select().where(cls.org==org).order_by(cls.id.asc())
+
+        if groups:
+            notification_destinations = notification_destinations.join(NotificationDestinationGroup).where(NotificationDestinationGroup.group << groups)
+
+        return notification_destinations
+
+    @property
+    def groups(self):
+        groups = NotificationDestinationGroup.select().where(NotificationDestinationGroup.notification_destination==self)
+        return dict(map(lambda g: (g.group_id, g.view_only), groups))
+
+
+class NotificationDestinationGroup(BaseModel):
```

Comment:
Do we really need to manage per group permissions of notifications?

Maybe it will be simpler to have the following permission model:
1. Everyone can see all `NotificationDestination`s and use them.
2. Only the owner (the user who created it) and admins can edit them.

WDYT?